Task-WebSocket.md

```diff
@@ -0,0 +1,98 @@
+# SaltyRTC WebSocket Task
+
+This task uses the end-to-end encrypted WebSocket connection set up by
+the signalling channel to send user defined messages. Note that there is
+no handover to a separate signaling channel.
+
+# Conventions
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this
+document are to be interpreted as described in
+[RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+# Terminology
+
+All terms from the [SaltyRTC protocol
+specification](./Protocol.md#terminology) are valid in this document.
+
+# Task Protocol Name
+
+The following protocol name SHALL be used for task negotiation:
+
+`v0.websocket.tasks.saltyrtc.org`
+
+# Task Data
+
+This task does not currently make use of the *data* field of the 'auth'
+messages described in the [SaltyRTC protocol
+specification](./Protocol.md#auth-message). The task data sent by the
+peers should be `Nil`.
+
+# Message Structure
+
+The same message structure as defined in the [SaltyRTC protocol
+specification](./Protocol.md#message-structure) SHALL be used.
+
+For simplicity, even though the `source` and `destination` fields in the
+nonce are not necessary anymore, the none structure is left unchanged.
+
+## 'close' Message
+
+The message itself and the client's behaviour is described in the
+[SaltyRTC protocol specification](./Protocol.md#close-message).
+
+## 'application' Message
+
+This message SHALL NOT be used in the WebSocket task.
+
+## User Defined Messages
+
+The user MAY send any MessagePack encoded message, as long as it follows
+the following format:
+
+* The message MUST be a `Map`
+* The message MUST contain a *type* key with a string value
+* The type SHALL NOT be 'application' or 'close'
```

Comment:
Maybe
> SHALL NOT be any of the message types defined by the SaltyRTC Signalling Protocol

Reply:
In theory types like `key` are fine, the client should know from the handshake state that these are user defined messages and not handshake messages.

Reply:
I'd like to avoid them. It might make sense to just say every type needs to be prefixed with `data-`.

Reply:
That adds some overhead that we tried to avoid. If we want some type of differentiator, then using a separate message type would probably be cleaner.

```
{
  "type": "data",
  "d": any
}
```

But I'm not sure we want those additional bytes in every message if it's not necessary.

Reply:
I think reusing types is awkward and should be avoided at all cost. It's also much uglier to debug. Your last proposal is definitely the cleanest way to solve this (although I'd say rename *d* to *data*) and I'm pretty sure it will be sufficient and performant enough for most use cases.